src/main.ts

```diff
@@ -978,6 +1103,172 @@ function generateEncode(ctx: Context, fullName: string, messageDesc: DescriptorP
   return joinCode(chunks, { on: '\n' });
 }
 
+/** Creates a function to encode a message by loop overing the tags then delimits the encoding with a length prefix */
+function generateEncodeDelimited(ctx: Context, fullName: string, messageDesc: DescriptorProto): Code {
+  const { options, utils, typeMap } = ctx;
+  const chunks: Code[] = [];
+
+  // create the basic function declaration
+  chunks.push(code`
+    encodeDelimited(
+      ${messageDesc.field.length > 0 ? 'message' : '_'}: ${fullName},
+      writer: ${Writer} = ${Writer}.create(),
+    ): ${Writer} {
```

Comment:
Same musing, can `encodeDelimited` just call `encode` after writing the prefix? Also, is the prefix being written? I'm scanning it but don't see it right now.

Reply:
@AlaaZorkane could you clarify how the length delimiter is being written before calling encode  here?

I am not familiar with the encoding logic, but I understand it as follows:

- `.ldelim()` writes a length-prefix, and then writes the message content _after_ it? So basically `message.ldelim()` outputs `[length, message]`
- encode itself already calls `ldelim()`, so it writes its own length delimiter
- you're calling `encode().ldelim()`, so that means an _extra_ delimiter is inserted.

Questions:
- Is this description correct?
- Do we then not write 2 delimiters each time we encode?
- If encoding already outputs its own length prefix _before_, why do we need to delimit messages at all?

Reply:
@boukeversteegh Of course! AFAIK:

- Yes an extra delimiter for the **whole** message as in `[length, message]` which in details can be per example: `[length, tag, length, data, tag, length, data...]`
- We do, but one is for the last type/value in a message (which includes a lot of types/values) not for the **whole** message.
- Protocol buffers are not length delimited by default, they length delimits encoded values inside of them, think of use cases where you have the value `42` encoded as `int64`, pb won't encode all `8 bytes` (I think there is an exception for `fixed` scalars but I never used them before) - to know how many bytes are encoded we must prefix the specific value `42` with a length

I made this little drawing that's missing a lot of stuff but kinda explains what I'm trying to say:

![image](https://user-images.githubusercontent.com/2952507/148290050-cf2031d3-e25a-432e-b939-3d9aa7322a80.png)


This is a good article to read about delimiting **whole** messages for streaming purposes: https://sureshjoshi.com/development/streaming-protocol-buffers

Reply:
Thank for your explanation @AlaaZorkane . It makes it more clear now!